client/network/sync/src/block_request_handler.rs

```diff
@@ -469,3 +490,122 @@ enum HandleRequestError {
 	#[error("Failed to send response.")]
 	SendResponse,
 }
+
+pub struct FullBlockDownloader {
+	protocol_name: ProtocolName,
+}
+
+impl FullBlockDownloader {
+	fn new(protocol_name: ProtocolName) -> Self {
+		Self { protocol_name }
+	}
+
+	/// Extracts the blocks from the response schema.
+	fn blocks_from_schema<B: BlockT>(
+		&self,
+		request: &BlockRequest<B>,
+		response: BlockResponseSchema,
+	) -> Result<Vec<BlockData<B>>, String> {
+		response
+			.blocks
+			.into_iter()
+			.map(|block_data| {
+				Ok(BlockData::<B> {
+					hash: Decode::decode(&mut block_data.hash.as_ref())?,
+					header: if !block_data.header.is_empty() {
```

Comment:
This whole code could be cleaned up with `then_some()`

Reply:
I made the change which caused tests to fail. The eager evaluation of `then_some()` turned out to be the issue(took some binary search to figure it out). The APIs like `decode_all()` get called even when they should not be (e.g) empty `vec` and return error, etc causing the block extraction to fail. Using `then()` is not so ergonomic in this context, so I rolled back the change.